### ISSUE TITLE
document uuid functions

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#12507]
`Base.Random` defines a `UUID` type and a `uuid4()` function to generate random [UUIDs](https://en.wikipedia.org/wiki/Universally_unique_identifier) (see also #12506) corresponding to [RFC 4122](http://www.ietf.org/rfc/rfc4122.txt), but they don't seem to be documented.  Probably we should document them.
- They are already used in several packages (IJulia, Bokeh, Escher, Jewel, and ProfileView)
- There is a redundant external package for this (forio/UUID.jl#13)
- Time-based "UUIDs" are employed in the core Julia to timestamp modules (https://github.com/JuliaLang/julia/blob/dcc448b89c3d743489cbce0bff881fd3b7799597/src/module.c#L32); we should maybe use a proper UUID algorithm for this.

**Comment by JeffBezanson** [comment]
The UUID.jl package seems to be more thorough. It's not that much code, so if we're going to use UUIDs in Base then we should probably move the package into Base and replace `Base.uuid4`.

**Comment by StefanKarpinski** [comment]
Packages will also need UUIDs in order to make renaming them work smoothly.

**Comment by stevengj** [comment]
The UUID.jl package is certainly more verbose.  The implementation of `uuid4` in base seems far cleaner, though, and makes me think that `uuid1` could be cleaned up similarly.

**Comment by WestleyArgentum** [comment]
Huh, `Base.uuid4` is definitely a lot cleaner... is it to spec though? Think it's RFC 4122... if no one is quite sure I could probably find some time this weekend to read through it and port UUID.jl into base

**Comment by wildart** [comment]
Please, put UUID functionality into Base.

**Comment by StefanKarpinski** [comment]
Relevant portion of the spec:

https://tools.ietf.org/html/rfc4122#section-4.4

I think it's correct but could probably use a double check.

**Comment by WestleyArgentum** [comment]
Yep, double checked, pretty cool. Seems like v1 is going to be a little more complicated, but I'll see if I can implement that and some of the conversions and things we have in UUID.jl

**Comment by stevengj** [comment]
v1 doesn't seem too bad, something like:
```
function uuid1(rng::AbstractRNG=Base.Random.GLOBAL_RNG)
    timestamp = (UInt64(now() - DateTime(1582,10,15))*10) | 0x1000000000000000
    clocksequence = rand(rng, UInt16) & 0x3fff
    node_id = (rand(rng, UInt64) & 0x0000ffffffffffff) | 0x0000010000000000
    return Base.Random.UUID(#= ...combine as specified in 4.1.2... =#)
end
```

**Comment by stevengj** [comment]
Or rather, `now(Dates.UTC)`.

**Comment by WestleyArgentum** [comment]
Lol, sorry, stepped out to dinner. Will check when back but that looks just
about right. The one thing I'd say is that we should accept a clock seq,
default it, and call random. And I remember seeing the python
implementation saved a global last timestamp that it would increment if u
got the same value because u were running fast or something

**Comment by stevengj** [comment]
Better, yet, generate `rand(rng, UInt128)` as in `uuid4()` and then mask out the bits for the timestamp, version, etc., to be or-ed in.

**Comment by stevengj** [comment]
@WestleyArgentum, my understanding was that the whole point of the random "clock sequence" was to handle the case of repeated timestamps, so I don't think a global is necessary.

**Comment by WestleyArgentum** [comment]
Re: clock sequence, looks like it's intended as a potential optimization
(if u know the previous then all u have to do is inc), but if we wrote the
algorithm to first generate a 128bit random number and masked things then
there's no additional overhead, so we could omit it. Then the only time u
might want it is if you're using it to keep track of ids or something, but
that seems extremely unlikely.

**Comment by WestleyArgentum** [comment]
I don't think this is quite right yet, but this is what I'm thinking:

```
function uuid1(rng::AbstractRNG=Base.Random.GLOBAL_RNG)
    u = rand(rng, Uint128)

    # mask off clock sequence and node
    u &= 0x00000000000000003fffffffffffffff

    # set the unicast/multicast bit
    u |= 0x00000000000000000000010000000000

    # 0x01b21dd213814000 is the number of milliseconds between
    # the UUID epoch and Unix epoch
    timestamp = round(Uint64, time() * 1e7) + 0x01b21dd213814000
    ts_low = timestamp & typemax(Uint32)
    ts_mid = (timestamp >> 32) & typemax(Uint16)
    ts_hi_and_ver = (timestamp >> 48) & 0x0fff

    u |= convert(Uint128, ts_low) << 96
    u |= convert(Uint128, ts_mid) << 80
    u |= convert(Uint128, ts_hi_and_ver) << 64

    Base.Random.UUID(u)
end
```

@stevengj I like how clean your timestamp calculation is, but I didn't quite understand why you were bitwise or-ing with `0x1000000000000000` and I wasn't sure about the resolution of `now(Dates.UTC)`. Am I being dumb?

**Comment by stevengj** [comment]
Now that we've merged UUID1 etc., we should think about documenting exporting.  Any objections to simply exporting `uuid1()` and `uuid4()`?

**Comment by WestleyArgentum** [comment]
It's definitely less used, but we should at least include documentation for (and maybe also export) `uuid_version(u::UUID)`. I've used it in past when looking at datasets that included mysterious uuids, and I'm sure at some point someone will be grateful to discover it alongside `uuid1` and `uuid4` before they've gone and re-implemented it.

**Comment by StefanKarpinski** [comment]
Would be a really nice place for `.` overloading: `uuid.version` instead of `uuid_version(uuid)`.

**Comment by ssfrr** [comment]
why not `version(u::UUID)`? Dispatch FTW or am I missing something?

**Comment by StefanKarpinski** [comment]
It's not clear to me that it makes sense to have a generic function to get the version of things. I.e. I don't think that's a universal concept, so having a shared function for it is inherently a pun. In this case, for example, you'd want to return an `Int`, but in other cases (e.g. `version(pkg)`) you'd probably want to return a `VersionNumber` object, so there's no hope of writing generic code using this function. Moreover, the usage of the term "version" for UUID variants is arguably a misuse of the term.

**Comment by WestleyArgentum** [comment]
I like `version(u::UUID)`, but I was afraid it was too "general"... hadn't considered `uuid.version` (didn't realize I could, to be honest, but I like the idea)

**Comment by ScottPJones** [comment]
@StefanKarpinski Any hope of `.` overloading (hopefully with `.!` instead of `..` for the "real access" operator) in 0.5?  That would help on so many fronts (to guarantee invariants of mutable types is my biggest need for them)

**Comment by StefanKarpinski** [comment]
It's possible. @JeffBezanson has some concerns that need to be overcome. Then there's the matter of implementation. It would probably be more likely with a prototype implementation to start from.

**Comment by ssfrr** [comment]
@StefanKarpinski that's an angle on julian style that I hadn't considered. One thing I really like about dispatch is that you can define and export a short and clear name on your type and it doesn't pollute the namespace (in the sense of "using up" a name). It seems like a shame not to use the shorter name if it's unambiguous, even though user code won't use it for generic programming.

Obviously I see a fraction of the volume of Julia code in the wild as you do so I'm maybe missing cases where this causes problems, but I always felt that dispatch created a low barrier to creating short, readable, unqualified names that lets us escape from methods_with_type_names_in_them. I'm curious as to the downsides.

**Comment by stevengj** [comment]
People, please take discussion of dot overloading to #1974 (if you have anything new to say); see also #5848 for implementation work.

**Comment by StefanKarpinski** [comment]
Yes, this definitely needs docs as well as filling in the other `uuidN` functions.

**Comment by matbesancon** [comment]
I'm preparing a PR, the docs are ready, I'll try to implement the uuid2/3

**Comment by matbesancon** [comment]
should this issue be closed, since it focuses on documentation of UUID types and functions, which has been addressed in #28649 
uuid3 don't seem feasible without MD5 hashing in Base, uuid2 is not so common

**Comment by stevengj** [comment]
Yes, I think this can be closed now that we have a UUIDs module in the standard library that exports `uuid4` etcetera, which now have docstrings.
